src/cpu/flexcpu/simple_dataflow_thread.hh

```diff
@@ -0,0 +1,620 @@
+     * execution of an instruction for which we previously requested execution.
+     *
+     * This function should be called after a call to executeInstruction() for
+     * the corresponding InflightInst, which should make a call to
+     * requestExecution() on the CPU. The order in which these results come in
+     * may not be obvious, since their execution may be dependent on scheduling
+     * or structural hazards, depending on the CPU implementation. They are
+     * guaranteed to occur AFTER the request is sent, at least.
+     */
+    void onExecutionCompleted(std::weak_ptr<InflightInst> inst, Fault fault);
+
+    // May want a different onCompleted for memory requests, if we someday want
+    // to have different behavior, but right now we don't so we just let
+    // onExecutionCompleted handle both types of instructions.
+
+    /**
+     * This function serves as the event handler for when the CPU has provided
+     * us with the data we requested to be fetched.
+     *
+     * This function should be called after onPCTranslated().
+     */
+    void onInstDataFetched(std::weak_ptr<InflightInst> inst,
+                              const TheISA::MachInst fetch_data);
+
+    /**
+     * This function serves as the event handler for when the CPU has completed
+     * the translation we requested for fetching purposes.
+     *
+     * This function should be called after onBeginFetch().
+     *
+     * @param fault If the translation request generated a fault, it should be
+     *  returned through this parameter.
+     * @param req A reference to a request object which should hold the
+     *  translated physical address.
+     */
+    void onPCTranslated(std::weak_ptr<InflightInst> inst, Fault fault,
+                        const RequestPtr& req);
+
+    bool outstandingFault()
+    { return earliestOutstandingFaultingInstNum > lastCommittedInstNum; }
+
+    /**
+     * This function detects any dependencies that a new instruction has, if it
+     * were to be issued at the current moment, considering the state of all
+     * currently active InflightInsts. The dependencies should be added to this
+     * instruction via its addDependency() interface function.
+     *
+     * @param inst The instruction for which we want to detect dependencies.
+     */
+    void populateDependencies(std::shared_ptr<InflightInst> inst);
+
+    /**
+     * This function iterates through the registers that this instruction
+     * writes and associates the registers with the corresponding instructions
+     * in order to streamline dependency tracking. Like a scoreboard.
+     *
+     * @param inst The instruction for which we want to track uses.
+     */
+    void populateUses(std::shared_ptr<InflightInst> inst);
+
+    /**
+     * This function checks if the thread should be active at this moment, and
+     * whether or not we have a reasonable PC value ready to be able to
+     * initiate another fetch at this time.
```

Comment:
Could you explain why you need this function (i.e., to deal with control flow instructions)